### PR TITLE
Auto memory management for MKLDNN

### PR DESCRIPTION
**Pull request by Menooker**
## What changes were proposed in this pull request?

Add auto memory management for MKLDNN native memory. Added a new trait `MemoryOwner`, which tracks all MKLDNN resources that needs to be released. The scala objects that may own any MKLDNN memory should extend this trait. For all MKLDNN memory allocation functions, an implicit parameter of type `MemoryOwner` is added. The allocated memory will be registered in the `MemoryOwner`. The `releaseNativeMklDnnMemory` should be called to release the resources.

## How was this patch tested?

UT

### CODE REVIEW

**Comment by wzhongyuan** [comment]
@i8run  @zhangxiaoli73   please take a look

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryOwner.scala [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.intel.analytics.bigdl.nn.mkldnn
+
+import com.intel.analytics.bigdl.tensor.DnnTensor
+import scala.collection.mutable.ArrayBuffer
+
+trait MemoryOwner {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryOwner.scala]: private[bigdl] ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala [hunk 2]
```diff
@@ -293,6 +297,7 @@ trait MklDnnLayer extends AbstractModule[Activity, Activity, Float] with MklDnnM
     }
 
     tensors.foreach(_.release())
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala]: useless

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryOwner.scala [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+  @transient
+  private lazy val _tensorMMaps: ArrayBuffer[TensorMMap] =
+    new ArrayBuffer[TensorMMap]()
+
+  @transient
+  private lazy val _tensors: ArrayBuffer[DnnTensor[_]] =
+    new ArrayBuffer[DnnTensor[_]]()
+
+  @transient
+  private  var _reorderManager: ReorderManager = _
```
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryOwner.scala]: replace by ReorderMemory

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryOwner.scala (hunk 1 above)
Comment (zhangxiaoli73) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/MemoryOwner.scala]: Add some comments

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala [hunk 4]
```diff
@@ -26,7 +26,7 @@ import scala.collection.mutable.ArrayBuffer
 /**
  * Helper utilities when integrating Module with MKL-DNN
  */
-trait MklDnnModule extends MklDnnModuleHelper {
+trait MklDnnModule extends MklDnnModuleHelper  {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala]: redundant space ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala [hunk 5]
```diff
@@ -78,7 +78,10 @@ trait MklDnnModule extends MklDnnModuleHelper {
   def setQuantize(value: Boolean): this.type
 }
 
-trait MklDnnModuleHelper {
+trait MklDnnModuleHelper extends MemoryOwner {
+
+  @transient protected implicit lazy val _this : MemoryOwner = this
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala]: _memoryOwner ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala [hunk 6]
```diff
@@ -123,7 +126,7 @@ trait MklDnnModuleHelper {
   }
 }
 
-trait MklDnnLayer extends AbstractModule[Activity, Activity, Float] with MklDnnModule {
+trait MklDnnLayer extends AbstractModule[Activity, Activity, Float] with MklDnnModule{
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala]: space

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala [hunk 7]
```diff
@@ -80,7 +80,7 @@ trait MklDnnModule extends MklDnnModuleHelper  {
 
 trait MklDnnModuleHelper extends MemoryOwner {
 
-  @transient protected implicit lazy val _this : MemoryOwner = this
+  @transient protected implicit lazy val _implicitMemoryOwner : MemoryOwner = this
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/DnnBase.scala]: why add implicit ?

**Comment by zhangxiaoli73** [comment]
LGTM

**Comment by Menooker** [comment]
Jenkins pass: http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/862/

**Pull request merged**